File: javascript/minesweeper/minesweeper.js
//
// This is only a SKELETON file for the 'Minesweeper' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const annotate = (input) => {
  const matrix = []

  for (const [rowPos, row] of Object.entries(input)) {
    const points = row.split('')
    input[rowPos] = points
  }
  
  for (const [rowPos, row] of Object.entries(input)) {
    matrix[rowPos] = []

    for (const [colPos, point] of Object.entries(row)) {
      if (point === "*") {
        matrix[rowPos][colPos] = "*"
        continue
      }

      let bombsAround = 0
      const directions = [
        [-1, -1], [-1, 0], [0, -1],
        [1, 1], [1, 0], [0, 1],
        [-1, 1], [1, -1]
      ]

      for (const dir of directions) {
        const newRowPos = dir[0] + parseInt(rowPos)
        const newColPos = dir[1] + parseInt(colPos)
        console.log(newRowPos, newColPos, newRowPos in input)
        if (newRowPos in input) {
          console.log(newColPos in input[newRowPos])
        }

        if (newRowPos in input &&
            newColPos in input[newRowPos] &&
            input[newRowPos][newColPos] === "*") {
          bombsAround++
        }
      }

      matrix[rowPos][colPos] = bombsAround === 0 ? " " : bombsAround
    }

    console.log(matrix[rowPos].join(''))
    matrix[rowPos] = matrix[rowPos].join('')
  }

  return matrix
}

File: python/minesweeper/minesweeper.py
def annotate(minefield):
    # Matrix of m[row][col] = v
    minefield = [[point for point in points] for points in minefield]

    # sanity pass, row must be consistently long
    length = -1
    for row in minefield:
        if length == -1:
            length = len(row)

        if length != len(row):
            raise ValueError('The board is invalid with current input.')
            

    for i, row in enumerate(minefield):
        for j, col in enumerate(row):
            if col != ' ' and col != '*':
                raise ValueError("The board is invalid with current input.")

            if col == '*':
                continue

            directions = [
                (-1, -1), (-1, 0), (0, -1),
                (1, 1), (1, 0), (0, 1),
                (-1, 1), (1, -1)
            ]
            bombsAround = 0

            for direction in directions:
                newRow = direction[0] + i
                newCol = direction[1] + j

                if 0 <= newRow < len(minefield) and \
                   0 <= newCol < len(minefield[newRow]) and \
                   minefield[newRow][newCol] == '*':
                    bombsAround += 1

            minefield[i][j] = bombsAround if bombsAround > 0 else ' '

        minefield[i] = "".join(map(str, minefield[i]))
    return minefield
